segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -652,14 +654,22 @@ private void triggerCacheTailIndex(DirectSegmentAccess segment, long tailCaching
         // Begin a loop which will end when either we've reached the target offset (defined in the RecoveryTask) or when
         // the recovery task itself has been cancelled (i.e., segment evicted, shutting down, etc.).
         Map<UUID, Long> tailCachePreIndexVersionTracker = new ConcurrentHashMap<>();
+        TailUpdates updates = new TailUpdates();
         Futures.loop(
                 () -> !task.task.isDone() && preIndexOffset.get() < task.triggerIndexOffset,
                 () -> {
                     int maxLength = (int) Math.min(this.config.getMaxTailCachePreIndexBatchLength(), task.triggerIndexOffset - preIndexOffset.get());
-                    return preIndexBatch(segment, preIndexOffset.get(), maxLength, tailCachePreIndexVersionTracker, lastIndexedOffset)
+                    return preIndexBatch(segment, preIndexOffset.get(), maxLength, tailCachePreIndexVersionTracker, lastIndexedOffset, updates)
```

Comment:
I am not clear that there is benefit to keeping the batching logic. If we are holding the whole of the data to be inserted into the cache in a local datastructure it seems like we might as well simplify it and just read it all in one go.

Reply:
`AsyncReadResultProcessor.processAll()` used internally in this method returns a `BufferView` with the actual Segment contents. I think that the reason for batching is just to limit the amount of heap memory being consumed at a given time. If we had to recover 1GB of unindexed data, it would imply to hold that much in memory while it is being processed. And this memory cost is per Table Segment, which there could be several per Segment Container.